snooty/semanticparser.py

```diff
@@ -48,24 +60,56 @@ def run(self, pages: Dict[FileId, Page]) -> Dict[str, SerializableType]:
             }
         )
 
-        return document
+        return document, self.diagnostics
 
     def run_event_parser(self) -> Dict[str, SerializableType]:
         event_parser = EventParser()
         event_parser.add_event_listener(OBJECT_START_EVENT, self.populate_include_nodes)
         event_parser.add_event_listener(
             OBJECT_START_EVENT, self.build_slug_title_mapping
         )
+        event_parser.add_event_listener(OBJECT_START_EVENT, self.validate_ref_targets)
         event_parser.consume(self.pages)
 
         # Return dict containing fields updated in event-based parse
         return {"slugToTitle": self.slug_title_mapping}
 
+    def validate_ref_targets(
+        self,
+        filename: FileId,
+        *args: SerializableType,
+        **kwargs: Optional[Dict[str, SerializableType]],
+    ) -> None:
+        """When a node of type ref_role is encountered, ensure that it references a valid target.
+
+        If so, append the full URL to the AST node. If not, throw an error.
+        """
+        obj = kwargs.get("obj")
+        assert obj is not None
+
+        if obj.get("type") == "ref_role":
+            domain = obj.get("domain")
+            name = obj.get("name")
+            target = obj.get("target")
+            assert all([isinstance(var, str) for var in [domain, name, target]])
+
+            key = f"{domain}:{name}:{target}"
+            if key not in self.targets:
```

Comment:
Python generally encourages you to ask for forgiveness rather than permission. e.g. this might be preferable:
```
try:
    # Append URL to AST
    obj["url"] = self.targets.get_url(key)
except KeyError:
    ...
```